DD_Plug.md

```diff
@@ -0,0 +1,153 @@
+    end
+  end
+end
+```
+
+Notice how just a few steps of authentication and authorization require complicated nested and duplication? Let's solve it with a couple plugs:
+
+```elixir
+defmodule HelloPhoenix.MessageController do
+  use HelloPhoenix.Web, :controller
+
+  plug :authenticate
+  plug :find_message
+  plug :authorize_message
+  plug :action
+  
+  def show(conn, params) do
+    render conn, :show, page: find_page(params["id"])
+  end
+  
+  defp authenticate(conn, _) do
+    case Authenticator.find_user(conn) do
+      {:ok, user} ->
+        assign(conn, :user, user)
+      :error ->
+        conn |> put_flash("You must be logged in") |> redirect(to: "/") |> halt
+    end
+  end
+
+  defp find_message(conn, _) do  
+    case find_message(params["id"]) do
+      nil -> 
+        conn |> put_flash("That message wasn't found") |> redirect(to: "/") |> halt
+      message ->
+        assign(conn, :message, message)
+    end
+  end
+  
+  defp authorize_message(conn, _) do
+    if Authorizer.can_acces?(conn.assigns[:user], conn.assigns[:message]) do
+      conn
+    else
+      conn |> put_flash("You can't access that page") |> redirect(to: "/") |> halt
+    end
+  end
+end
+```
+
+By replacing the nested blocks of code with a flattened series of plug transformations, we were able to take almost the identical code and make it much more composable, clear, and resuable.
+
+Now, let's look at the other flavor plugs come in: Module Plugs.
+
+### Module Plugs
+
+Module plugs are another type of Plug that lets you group a connection transformation into a module. The module only needs to implement two functions:
+
+- `init/1` initializes any arguments or options passed to be pased to call/2
+- `call/2` carries out the connectiont transformation. `call/2` is just a function plug that we saw earlier
+
+
+To see this in action, lets write a Module Plug that that puts the `:locale` assign into the connectiona assigns for downstream use in other plugs, the controller actions, and our views.
+
+```elixir
+defmodule HelloPhoenix.Plugs.Locale
+  import Plug.Conn
+  
+  @locales ["en", "fr", "de"]
+  
+  def init(default), do: default
+
+  def call(%Plug.Conn{params: %{"locale" => loc}}, _default) when loc in @locales
+    assign(conn, :locale, loc))
+  end
+  def call(conn, default), do: assign(conn, :locale, default)
+end
+
+defmodule HelloPhoenix.Router do
+  use Phoenix.Router
+
+  pipeline :browser do
+    plug :accepts, ~w(html)
+    plug :fetch_session
+    plug :fetch_flash
+    plug :protect_from_forgery
+    plug HelloPhoenix.Plugs.Locale, "en"
+  end
+  ...
+```
+
+By using a module plug, we were able to add it to our browser pipeline via `plug HelloPhoenix.Plugs.Locale, "en"`. We can see that using the `init/1` callback, we passed a default locale to fallback to if none is present. We also made use of pattern matching to define multiple `call/2` function heads to validate the locale in the params, and otherwise fallback to "en". 
```

Comment:
How about present tense here as well, ". . . we are able to add. . ."

And "We can see that by using. . . "